nsis/src/nsis/NsisAnalyzer.java

```diff
@@ -23,126 +21,166 @@
+            .getAddressString(/* display mnemonic */ true, /* pad address if necessary */ true));
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Disassembles the byte code in the specified memory block.
+   * 
+   * @param program to instanciate the disassembler with
+   * @param memoryBlock to perform the disassembly on
+   * @param monitor the TaskMonitor object to monitor the operation
+   * @return the AddressSet of the disassembled instructions
+   */
+  private AddressSet disassembleByteCode(Program program, MemoryBlock memoryBlock,
+      TaskMonitor monitor) {
+    Disassembler disassembler =
+        Disassembler.getDisassembler(program, monitor, /* Object to notify */ null);
+    AddressSet entriesAddrSet = new AddressSet(memoryBlock.getStart(), memoryBlock.getEnd());
+    return disassembler.disassemble(entriesAddrSet.getMinAddress(), entriesAddrSet,
+        /* follow flow */ true);
+  }
+
+  /**
+   * Resolve strings for the specified instruction.
+   * 
+   * @param instr the instruction
+   * @param stringsBlock the memory block containing the strings
+   * @throws MemoryAccessException
+   */
+  private void resolveStrings(Instruction instr, MemoryBlock stringsBlock)
+      throws MemoryAccessException {
+    String mnemonic = instr.getMnemonicString();
+    switch (mnemonic) {
+      case "MessageBox":
+        Address parameterAddr =
+            stringsBlock.getStart().add(instr.getInt(NsisConstants.ARG2_OFFSET));
+        instr.addOperandReference(NsisConstants.ARG2_INDEX, parameterAddr, RefType.PARAM,
+            SourceType.ANALYSIS);
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  /**
+   * Resolve the control flow for the specified instruction
+   * 
+   * @param instr the instruction
+   * @param entriesBlock the memory block containing the instructions
+   * @param program
+   * @throws MemoryAccessException
+   */
+  private void resolveControlFlow(Instruction instr, MemoryBlock entriesBlock,
+      ReferenceManager referenceManager) throws MemoryAccessException {
+    String mnemonic = instr.getMnemonicString();
+    int instructionNumber;
+    switch (mnemonic) {
+      case "Call":
+        instr.setFlowOverride(FlowOverride.CALL);
+        instructionNumber = instr.getInt(NsisConstants.ARG1_OFFSET);
+        referenceManager.addMemoryReference(instr.getAddress(),
+            entriesBlock.getStart()
+                .add((instructionNumber - 1) * NsisConstants.INSTRUCTION_BYTE_LENGTH),
+            RefType.UNCONDITIONAL_CALL, SourceType.ANALYSIS, NsisConstants.ARG1_INDEX);
+        break;
+
+      case "Jmp":
+        instr.setFlowOverride(FlowOverride.BRANCH);
+        instructionNumber = instr.getInt(NsisConstants.ARG1_OFFSET);
+        referenceManager.addMemoryReference(instr.getAddress(),
+            entriesBlock.getStart()
+                .add((instructionNumber - 1) * NsisConstants.INSTRUCTION_BYTE_LENGTH),
```

Comment:
I suggest that you make a small helper function (e.g. `getInstructionAddress`) for converting `instructionOffset` to `instructionAddress` and not do this calculation in every switch. That would also make the code more readable.

Reply:
I need to do that calculation in every switch, because it depends on the value of the parameter (instructionNumber), and this parameter is not always the first argument of every instruction. I'll create a helper function and I will call that function with instructionNumber for each instruction

Reply:
Yes, that's exactly what I meant.